src/codegen/llvm/codegen_llvm_helper_visitor.cpp

```diff
@@ -534,17 +529,43 @@ void CodegenLLVMHelperVisitor::visit_nrn_state_block(ast::NrnStateBlock& node) {
     /// convert local statement to codegenvar statement
     convert_local_statement(*loop_block);
 
-    /// create for loop node
-    auto for_loop_statement = std::make_shared<ast::CodegenForStatement>(initialization,
-                                                                         condition,
-                                                                         increment,
-                                                                         loop_block);
+    /// main loop possibly vectorized on vector_width
+    {
+        /// loop constructs : initialization, condition and increment
+        const auto& initialization = create_statement_as_expression("id = 0");
+        const auto& condition = create_expression("id < node_count");
+        const auto& increment = create_statement_as_expression("id = id + {}"_format(vector_width));
+
+        auto for_loop_statement_main = std::make_shared<ast::CodegenForStatement>(initialization,
+                                                                                  condition,
+                                                                                  increment,
+                                                                                  loop_block);
+
+        /// convert all variables inside loop body to instance variables
+        convert_to_instance_variable(*for_loop_statement_main, loop_index_var);
+
+        /// loop itself becomes one of the statement in the function
+        function_statements.push_back(for_loop_statement_main);
+    }
+
+    /// remainder loop possibly vectorized on vector_width
+    {
+        /// loop constructs : initialization, condition and increment
+        const auto& initialization = create_statement_as_expression("");
+        const auto& condition = create_expression("id < node_count");
+        const auto& increment = create_statement_as_expression("++id");
 
-    /// convert all variables inside loop body to instance variables
-    convert_to_instance_variable(*for_loop_statement, loop_index_var);
+        auto for_loop_statement_remainder = std::make_shared<ast::CodegenForStatement>(initialization,
+                                                                                  condition,
+                                                                                  increment,
+                                                                                  loop_block);
```

Comment:
`loop_block` is a pointer and hence we need to clone the loop_block (to avoid side effects - if we transform first loop then 2nd will be also modified)

Reply:
Don't we want that both loop are exactly the same?

Reply:
No. Vector loop can have extra transformations:

```
# original loop
for(....) {
  int x
  ...
}

# will become

for (...) {
  int x<8>  // width of 8
  ...
}

for (....) { // reminder 
  int x       => this shouldn't be changed!
  ...
}
```